lib/react.dart

```diff
@@ -775,6 +843,43 @@ abstract class Component2 implements Component {
   List<StateUpdaterCallback> get transactionalSetStateCallbacks => throw new UnimplementedError();
 }
 
+/// Mixin that enforces consistent typing of the `snapshot` parameter
+/// returned by [Component2.getSnapshotBeforeUpdate]
+/// and passed into [Component2.componentDidUpdate].
+///
+/// __EXAMPLE__:
+///
+/// class _ParentComponent extends react.Component2
+///     with react.TypedSnapshot<bool> {
+///
+///   // Note that the return value is specified as an bool, matching the mixin
+///   // type.
+///   bool getShapshotBeforeUpdate(prevProps, prevState) {
+///     if (prevProps["foo"]) {
+///       return true;
+///     }
+///     return null;
+///   }
+///
+///   // Note that the snapshot parameter is consistently typed with the return
+///   // value of getSnapshotBeforeUpdate.
+///   void componentDidUpdate(prevProps, prevState, [bool snapshot]) {
+///     if (snapshot == true) {
+///       // Do something
+///     }
+///   }
+///
+///   /* Include other standard component logic */
+///
+/// }
+abstract class TypedSnapshot<TSnapshot> implements Component2 {
```

Comment:
This is a perfect situation to utilize Dart 2's `mixin`s.  It will allow use to keep the methods abstract, and is cleaner / less confusing than `implements`, IMO.

```dart
mixin TypedSnapshot<TSnapshot> on Component2 {
  @override
  TSnapshot getSnapshotBeforeUpdate(Map prevProps, Map prevState);

  @override
  void componentDidUpdate(Map prevProps, Map prevState, [covariant TSnapshot snapshot]);
}
```

Reply:
Also - note the use of the `@override` annotation.  Its a good idea to get in the habit of annotating overridden members.